core/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -378,13 +381,26 @@ private void checkFieldUniqueness(String type, Collection<ObjectMapper> objectMa
                 throw new IllegalArgumentException("Field [" + name + "] is defined twice in [" + type + "]");
             }
         }
+
+        // then check other types
+        for (String fieldName : fieldNames) {
+            if (fullPathObjectMappers.containsKey(fieldName)) {
+                throw new IllegalArgumentException("[" + fieldName + "] is defined as a field in mapping [" + type
+                        + "] but this name is already used for an object in other types");
+            }
+        }
+
+        for (String objectPath : objectFullNames) {
+            if (fieldTypes.get(objectPath) != null) {
+                throw new IllegalArgumentException("[" + objectPath + "] is defined as an object in mapping [" + type
+                        + "] but this name is already used for a field in other types");
+            }
+        }
     }
 
     private void checkObjectsCompatibility(String type, Collection<ObjectMapper> objectMappers, Collection<FieldMapper> fieldMappers, boolean updateAllTypes) {
         assert Thread.holdsLock(this);
 
-        checkFieldUniqueness(type, objectMappers, fieldMappers);
```

Comment:
I removed it because we already call checkFieldUniqueness from merge(), which also callo checkObjectsCompatibility.